x/pricefeed/simulation/operations/msg.go

```diff
@@ -0,0 +1,149 @@
+)
+
+var (
+	noOpMsg     = simulation.NoOpMsg(pricefeed.ModuleName) // TODO QUESTION what is the purpose of noopmsg?
+	govDenom    = "ukava"                                  // TODO QUESTION do I need these in pricefeed?
+	stableDenom = "usdx"
+)
+
+// SimulateMsgUpdatePrices updates the prices of various assets
+// TODO - MUST include BNB, USDX, BTC as HARD REQUIREMENTS - asset prices needed by CDP sims
+func SimulateMsgUpdatePrices(authKeeper auth.AccountKeeper, keeper keeper.Keeper) simulation.Operation {
+	// get a pricefeed handler
+	handler := pricefeed.NewHandler(keeper)
+
+	// There's two error paths
+	// - return a OpMessage, but nil error - this will log a message but keep running the simulation
+	// - return an error - this will stop the simulation (I think)
+	return func(r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simulation.Account) (
+		simulation.OperationMsg, []simulation.FutureOperation, error) {
+
+		// OVERALL LOGIC IDEA:
+		// (1) RANDOMLY PICK AN ASSET OUT OF BNB AN BTC [TODO QUESTION - PRESUMABLY USDX IS EXCLUDED AS IT IS A STABLE DENOM?]
+		// (2) GET THE CURRENT PRICE OF THAT ASSET IN USD
+		// (3) GENERATE A RANDOM NUMBER IN THE RANGE 0.8-1.2 (UNIFORM DISTRIBUTION) [TODO POTENTIAL UPDATE MAKE IT NORMALLY DISTRIBUTED ACROSS A WIDER RANGE?]
+		// (4) MULTIPLY THE CURRENT PRICE BY THE RANDOM NUMBER
+		// (5) POST THE NEW PRICE TO THE KEEPER
+
+		// get the address for the account
+		// TODO QUESTION do we need to check if this address exists on chain or is an oracle for the price or does it not matter
+		address := sdk.AccAddress("kava1ffv7nhd3z6sych2qpqkk03ec6hzkmufy0r2s4c")
+
+		// pick a random asset out of BNB and BTC
+		assetCode, err := pickRandomAsset(r)
+		if err != nil {
+			return noOpMsg, nil, fmt.Errorf("Error picking random asset")
+		}
+
+		fmt.Printf("Picked asset: %s\n", assetCode)
+
+		// TODO QUESTION - GET THE CURRENT PRICE OF THE ASSET?? HOW TO DO THIS?
+		currentPrice := int64(100)
+
+		// generate a new random price based off the current price
+		price, err := pickNewRandomPrice(r, currentPrice)
+		if err != nil {
+			return noOpMsg, nil, fmt.Errorf("Error picking random price")
+		}
+
+		// get the expiry time based off the current time
+		expiry := getExpiryTime()
+
+		// MSG POST PRICE
+		// GENERATE THE MSG TO SEND TO THE KEEPER
+		// now create the msg to post price
+		msg := types.NewMsgPostPrice(address, assetCode, price, expiry)
+
+		// Perform basic validation of the msg - don't submit errors that fail ValidateBasic, use unit tests for testing ValidateBasic
+		if err := msg.ValidateBasic(); err != nil {
+			return noOpMsg, nil, fmt.Errorf("expected msg to pass ValidateBasic: %s", msg.GetSignBytes())
+		}
+
+		// now we submit the pricefeed update message
+		// TODO QUESTION - this is failing for some reason? Any ideas why?
+		if ok := submitMsg(ctx, handler, msg); !ok {
+			return noOpMsg, nil, fmt.Errorf("could not submit pricefeed msg")
```

Comment:
This is where the error is coming from @rhuairahrighairidh

Reply:
Had similar issues in bep3 module. Try the following:

1. Print handler log in `submitMsg()` i.e.
```
log := handler(ctx, msg).Log
fmt.Println(log)
```
2. Add print statements in `x/pricefeed/handler.go` before each `return err.Result()`

Reply:
You need to implement `x/pricefeed/simulation/genesis.go` before operations will work. In `genesis.go` you setup the genesis state (this determines which markets are in the pricefeed). You can see examples of pricefeed genesis state in `x/pricefeed/keeper/integration_test.go`.  I think having one genesis state for pricefeed is fine, so you don't need to randomize it. I would include xrp, btc, and bnb in the pricefeed genesis state (`params.Markets`). Further, you will need to dynamically modify the genesis state by adding oracles (account addresses) for each market and setting initial prices for each market. You can access accounts using `simState.Accounts`, so something like this in `RandomizedGenState`

```
// define getPricefeedSimulationParams to return params with xrp:usd, btc:usd, bnb:usd
params := getPricefeedSimulationParams()
genPrices := []PostedPrice{}
for _, market := range params.Markets {
  for _, acc := range simState.Accounts {
    market.Oracles = append(market.Oracles, acc.Address)
    // define getInitialPrice to set the initial price for each market (ie a switch statement where btc:usd is ~7000, bnb:usd is 14, and xrp:usd is 0.2(
    genPrice := types.PostedPrice{market.ID, acc.Address, getInitialPrice(market.ID), simState.GenTimestamp.Add(time.Hour * 24)}
    genPrices = append(genPrices, genPrice)
  }
}
pricefeedGenesis := types.NewGenesisState(params, genPrices)
fmt.Printf("Selected randomly generated %s parameters:\n%s\n", types.ModuleName, codec.MustMarshalJSONIndent(simState.Cdc, pricefeedGenesis))
simState.GenState[types.ModuleName] = simState.Cdc.MustMarshalJSON(pricefeedGenesis)
```

Reply:
ok